text/0000-source_replacement_ambiguity.md

```diff
@@ -0,0 +1,126 @@
+with alternative registries and ensures credentials are only sent to the registry they are
+associated with.
+
+### Change 4: `[source]` table can reference `[registries]` table
+The `replace-with` key in the `[source]` table can reference a registry defined in the `[registries]` table.
+
+For example, the following configuration would be valid:
+
+```
+[source.crates-io]
+replace-with = "my-registry"
+
+[registry.my-registry]
+index = "https://my-registry-index/"
+```
+
+This is necessary to allow the `--registry <NAME>` command-line argument to work with source-replaced registries. It also allows additional configuration (such as a token) to be specified for a source-replacement registry without duplicating configuration between `[registry]` and `[source]` tables.
+
+# Drawbacks
+[drawbacks]: #drawbacks
+
+Behavior is changed around where credentials are sent, which could break some workflows.
+
+If a mirror of crates.io is set up with `config.json` containing `"api": "https://crates.io"`, then the current system of sending the crates.io token to the replaced source would work correctly, and this RFC would break it.
+
+# Rationale and alternatives
+[rationale-and-alternatives]: #rationale-and-alternatives
+
+## Alternative: ignore source replacement for API operations
+When doing an API operation with a replaced `crates-io`, `cargo` would ignore source replacement without additional arguments. This is how alternative registries currently work.
+
+If the user wants to use the replacement, they could pass `--registry <NAME>`, but would not be required to do so.
+
+A new option `--respect-source-config` could be added to make cargo follow the source replacement for API operations (similar to what we already have for `cargo vendor`).
+
+This may be too confusing for users since it silently changes behavior. The RFC proposes a solution that requires the user to be explicit about which registry to use in the ambiguous situation (crates-io replacement).
+
+## Alternative: disallow source replacement for API operations
+
+Attempting an API operation on a replaced source would be an error. The user could use `--registry crates-io` to explicitly bypass the source replacement.
+```
+Error: <operation> is not supported on replaced source `crates-io-mirror`; use `--registry crates-io` for the original source
+```
+
+# Prior art
+[prior-art]: #prior-art
+
+Other package managers don't seem to have a source replacement feature.
+
+# Unresolved questions
+[unresolved-questions]: #unresolved-questions
+
+Should the `--registry <NAME>` command line argument be allowed to reference the name of a `source` from the `[source]` table as well? This makes it more flexible, but adds potentially unnecessary complexity.
+
+Cargo's tests rely on the ability to replace the crates.io source and have the crates.io credentials go to the replaced source. We need a way for these tests to continue working. 
```

Comment:
I'm not too worried about this, as I believe all of the relevant tests could be adjusted to include `--registry` if necessary. In the past, the tests would pass `--index`, though that was removed in https://github.com/rust-lang/cargo/pull/7973.

It is unfortunate that it is difficult to test cargo's behavior without using source replacement. There are some tests that are impossible to write, but I don't think it should be too much of a problem.